server/athenian/api/experiments/preloading/entries.py

```diff
@@ -1,8 +1,137 @@
-from athenian.api.controllers.features.entries import \
-    MetricEntriesCalculator as OriginalMetricEntriesCalculator
+from datetime import datetime
+from typing import Collection, Optional, Set, Tuple
+
+import aiomcache
+import databases
+import pandas as pd
+
+from athenian.api.controllers.features.entries import (
+    MetricEntriesCalculator as OriginalMetricEntriesCalculator,
+)
+from athenian.api.controllers.miners.filters import JIRAFilter, LabelFilter
+from athenian.api.controllers.miners.github.pull_request import (
+    PullRequestMiner as OriginalPullRequestMiner,
+)
+from athenian.api.controllers.miners.types import PRParticipants, PRParticipationKind
+from athenian.api.models.metadata.github import PullRequest
+from athenian.api.tracing import sentry_span
+
+
+class PullRequestMiner(OriginalPullRequestMiner):
+    """Load all the information related to Pull Requests from the metadata DB. Iterate over it \
+    to access individual PR objects."""
+
+    @classmethod
+    @sentry_span
+    async def _fetch_prs_by_filters(
+        cls,
+        time_from: datetime,
+        time_to: datetime,
+        repositories: Set[str],
+        participants: PRParticipants,
+        labels: LabelFilter,
+        jira: JIRAFilter,
+        exclude_inactive: bool,
+        pr_blacklist: Optional[Collection[str]],
+        pr_whitelist: Optional[Collection[str]],
+        meta_ids: Tuple[int, ...],
+        mdb: databases.Database,
+        cache: Optional[aiomcache.Client],
+        columns=PullRequest,
+        updated_min: Optional[datetime] = None,
+        updated_max: Optional[datetime] = None,
+    ) -> pd.DataFrame:
+        if jira or labels:
+            # TODO: not supported yet, call original implementation
+            return await super(PullRequestMiner, cls)._fetch_prs_by_filters(
+                time_from,
+                time_to,
+                repositories,
+                participants,
+                labels,
+                jira,
+                exclude_inactive,
+                pr_blacklist,
+                pr_whitelist,
+                meta_ids,
+                mdb,
+                cache,
+                columns=columns,
+                updated_min=updated_min,
+                updated_max=updated_max,
+            )
+
+        assert (updated_min is None) == (updated_max is None)
+
+        df = mdb.cache.dfs["prs"].df
+
+        mask = (
+            (~df["closed"] | (df["closed_at"] >= time_from))
```

Comment:
As a rule of thumb, try to avoid Pandas operations because they are much slower than plain numpy. Proved by numerous hours spent in the profiler. For example, here:
```python
(~df["closed"].values | (df["closed_at"].values >= np.array([time_from], dtype=df["closed_at"].dtype))
```
will work faster. There are some caveats with datetime64 though.

Reply:
Probably make sense to apply your changes for all operations except for those related to `datetime64`?

Reply:
Smth like this, yes. I'd say if it works with `np.datetime64` then awesome, if it doesn't - we have to fall back to Pandas.